src/fuse_attention.cpp

```diff
@@ -209,6 +214,395 @@ struct find_attention
     }
 };
 
+struct find_flash_decoding
+{
+    // number of groups. User-provided for now
+    std::size_t groups;
+
+    auto matcher() const
+    {
+        return match::name("group")(match::has_op_value("tag", "attention")).bind("group");
+    }
+
+    std::pair<instruction_ref, instruction_ref> get_gemms(module_ref submod) const
+    {
+        std::vector<instruction_ref> gemms;
+        for(auto it = submod->begin(); it != submod->end(); ++it)
+        {
+            if(it->name() == "dot")
+                gemms.push_back(it);
+        }
+        assert(gemms.size() == 2 and "Expected exactly 2 gemm operations in attention submodule");
+
+        // gemms[0] is Q@K, gemms[1] is P@V
+        // gemms are in order since we iterate from begin to end
+        return {gemms[0], gemms[1]};
+    }
+
+    std::vector<shape> get_qkv_shapes(instruction_ref q, instruction_ref k, instruction_ref v) const
+    {
+        std::vector<shape> qkv_shapes;
+        auto q_shape = q->get_shape();
+        auto k_shape = k->get_shape();
+        auto v_shape = v->get_shape();
+
+        qkv_shapes.push_back(q_shape);
+        qkv_shapes.push_back(k_shape);
+        qkv_shapes.push_back(v_shape);
+
+        assert((q_shape.lens().size() == 3 or q_shape.lens().size() == 4) and
+               "Expected 3D or 4D Q, K, V shapes");
+        assert(k_shape.lens().size() == q_shape.lens().size() and
+               v_shape.lens().size() == q_shape.lens().size() and
+               "Q, K, V must have same number of dimensions");
+        return qkv_shapes;
+    }
+
+    struct transformed_shapes_result
+    {
+        std::vector<size_t> q_shape;           // final Q shape: [B, G, M, k]
+        std::vector<size_t> k_intermediate;    // K intermediate: [B, k, G, N/G]
+        std::vector<size_t> k_shape;           // final K shape: [B, G, k, N/G]
+        std::vector<int64_t> k_transpose_perm; // permutation for K transpose
+        std::vector<size_t> v_shape;           // final V shape: [B, G, N/G, D]
+    };
+
+    transformed_shapes_result get_transformed_shapes(const std::vector<shape>& input_shapes) const
+    {
+        assert(input_shapes.size() == 3 and "Expected Q, K, V shapes");
+
+        auto q_lens = input_shapes[0].lens();
+        auto k_lens = input_shapes[1].lens();
+        auto v_lens = input_shapes[2].lens();
+
+        // 3D: Q_lens = [B, M, k]
+        // 4D: Q_lens = [B, H, M, k]
+        size_t ndim = q_lens.size();
+        size_t n    = k_lens[ndim - 1];
+        size_t g    = groups;
+
+        // TODO: do we wanna support this differently?
+        assert(n % g == 0 and
```

Comment:
might make sense to just not apply this flash decoding transformation in this case (ie. check this in the apply method and return). I assume you would need to add some sort of padding logic to support this case?